internal/pkg/api/handleUpload.go

```diff
@@ -0,0 +1,269 @@
+		writeUploadError(err, w, zlog, start, "error finalizing upload")
+		return
+	}
+}
+
+type UploadT struct {
+	bulker      bulk.Bulk
+	chunkClient *elasticsearch.Client
+	cache       cache.Cache
+	uploader    *upload.Uploader
+}
+
+func NewUploadT(cfg *config.Server, bulker bulk.Bulk, chunkClient *elasticsearch.Client, cache cache.Cache) *UploadT {
+	log.Info().
+		Interface("limits", cfg.Limits.ArtifactLimit).
+		Int64("maxFileSize", maxFileSize).
+		Msg("upload limits")
+
+	return &UploadT{
+		chunkClient: chunkClient,
+		bulker:      bulker,
+		cache:       cache,
+		uploader:    upload.New(chunkClient, bulker, maxFileSize, maxUploadTimer),
+	}
+}
+
+func (ut *UploadT) handleUploadStart(zlog *zerolog.Logger, w http.ResponseWriter, r *http.Request) error {
+	// decode early to match agentID in the payload
+	payload, err := upload.ReadDict(r.Body)
+	if err != nil {
+		if errors.Is(err, io.EOF) {
+			return fmt.Errorf("file info body is required: %w", err)
+		}
+		return err
+	}
+
+	// check API key matches payload agent ID
+	agentID, ok := payload.Str("agent_id")
+	if !ok || agentID == "" {
+		return errors.New("required field agent_id is missing")
+	}
+	if AUTH_ENABLED {
+		if _, err := authAgent(r, &agentID, ut.bulker, ut.cache); err != nil {
+			return err
+		}
+	}
+
+	// validate payload, enrich with additional fields, and write metadata doc to ES
+	info, err := ut.uploader.Begin(r.Context(), payload)
+	if err != nil {
+		return err
+	}
+
+	// prepare and write response
+	out, err := json.Marshal(map[string]interface{}{
+		"upload_id":  info.ID,
+		"chunk_size": info.ChunkSize,
+	})
+	if err != nil {
+		return err
+	}
+	w.Header().Set("Content-Type", "application/json")
+	_, err = w.Write(out)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (ut *UploadT) handleUploadChunk(zlog *zerolog.Logger, w http.ResponseWriter, r *http.Request, uplID string, chunkID int) error {
+	chunkHash := strings.TrimSpace(r.Header.Get("X-Chunk-Sha2"))
+	if chunkHash == "" {
+		return errors.New("chunk hash header required")
+	}
+
+	upinfo, chunkInfo, err := ut.uploader.Chunk(r.Context(), uplID, chunkID, chunkHash)
+	if err != nil {
+		return err
+	}
+
+	// prevent over-sized chunks
+	data := http.MaxBytesReader(w, r.Body, upload.MaxChunkSize)
+
+	// compute hash as we stream it
+	hash := sha256.New()
+	copier := io.TeeReader(data, hash)
+
+	ce := cbor.NewChunkWriter(copier, chunkInfo.Last, chunkInfo.BID, chunkInfo.SHA2, upinfo.ChunkSize)
+	if err := upload.IndexChunk(r.Context(), ut.chunkClient, ce, upinfo.Source, chunkInfo.BID, chunkInfo.Pos); err != nil {
+		return err
+	}
+
+	hashsum := hex.EncodeToString(hash.Sum(nil))
+
+	if !strings.EqualFold(chunkHash, hashsum) {
+		// delete document, since we wrote it, but the hash was invalid
+		// context scoped to allow this operation to finish even if client disconnects
+		if err := upload.DeleteChunk(context.Background(), ut.bulker, upinfo.Source, chunkInfo.BID, chunkInfo.Pos); err != nil {
```

Comment:
We should assume the hash is ok as it will be 99.9% of the time.  Writing to a temporary index implies a chunk rewrite on success, which we should avoid.